runtime/onert/api/include/nnfw.h

```diff
@@ -479,6 +498,19 @@ NNFW_STATUS nnfw_set_available_backends(nnfw_session *session, const char *backe
  */
 NNFW_STATUS nnfw_set_op_backend(nnfw_session *session, const char *op, const char *backend);
 
+/**
+ * @brief     Set the OpenCL Tuner Configuration
+ *
+ * This function should be called before {@link nnfw_prepare} is invoked.
+ *
+ *
+ * @param[in] session session to set OpenCL Tuner Configuration
+ * @param[in] cltuner cltuner to configure OpenCL Tuner Configuration Value
+ *
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_set_tuner(nnfw_session *session, nnfw_cltunerinfo *cltuner);
+
```

Comment:
IMHO, It's not good solution to add API for backend-specific feature.

Reply:
In case of other open source project, ARMNN, such backend specific information can be delivered by user to internal module using backend option[1]. So user can choose a inference device, CPU or GPU, and ARMNN provides the option which can pass GPU specific data to internal module with the target device. And looks reasonable to me.

Does ONERT provide such interface? If any then could you provide this committer the guide? If not any then I think you could guide a proper way for it rather than just say no.

AFAIK, ONE is an official open source project. So I'd kindly recommend you to give some help to this committer to find a proper solution. That would be one of very important factors for the growth of the community.


[1] https://github.com/ARM-software/armnn/blob/branches/armnn_20_11/include/armnn/IRuntime.hpp#L115

Reply:
>In case of other open source project, ARMNN, such backend specific information can be delivered by user to internal module using backend option[1]. So user can choose a inference device, CPU or GPU, and ARMNN provides the option which can pass GPU specific data to internal module with the target device. And looks reasonable to me.

I don't disagree to introduce general-type backend configuration API. My comment means that it looks acl-cl (or OpenCL) backend specific API.

>Does ONERT provide such interface? If any then could you provide this committer the guide? If not any then I think you could guide a proper way for it rather than just say no.

We don't have interface for backend configuration, so I recommand you to use `nnfw_explemental.h` (experimental API) instead of `nnfw.h` as temporary solution. I also don't have any solution to resolve this issue.

@wateret Could you have any idea about API (or interface for backend configuration) design?

Reply:
> I don't disagree to introduce general-type backend configuration API. My comment means that it looks acl-cl (or OpenCL) backend specific API.

I agree with this.

> I recommand you to use `nnfw_explemental.h`

And this is what I would recommend too 😄

Reply:
Eventually I would prefer introducing a way to pass some general data to backend.

What I just came up is something like below. However I'm not sure about this would be a good solution in many respects.

```c
// Pass data pointer to the backend context of `backend_id`
NNFW_STATUS nnfw_set_backend_data(nnfw_session *session, const char *backend_id, void *data);
```

Is there a good practice when we do this kind of things?

Reply:
What I am thinking is:

- The user(who uses NNFW API) knows what should go in `void *data`
- NNFW API Frontend and ONERT Core **knows nothing** but just pass the pointer to the ACL backend
- ACL backend knows the format of `void *data` and it got the pointer from Core, so make use of it

Reply:
> > In case of other open source project, ARMNN, such backend specific information can be delivered by user to internal module using backend option[1]. So user can choose a inference device, CPU or GPU, and ARMNN provides the option which can pass GPU specific data to internal module with the target device. And looks reasonable to me.
> 
> I don't disagree to introduce general-type backend configuration API. My comment means that it looks acl-cl (or OpenCL) backend specific API.
> 
> > Does ONERT provide such interface? If any then could you provide this committer the guide? If not any then I think you could guide a proper way for it rather than just say no.
> 
> We don't have interface for backend configuration, so I recommand you to use `nnfw_explemental.h` (experimental API) instead of `nnfw.h` as temporary solution. I also don't have any solution or idea to resolve this issue.
> 
> @wateret Could you have any idea about API (or interface for backend configuration) design?

@intom It may be a good way to introduce the interface for backend configuration instead of the API if ONERT has no any interface for it and other agree to this way. :) AFAIK, the backend configuration interface can be used for other purpose such as profiling and debugging.